File: src/search/binary_search.cpp
// Copyright (C) Leslie Zhai <xiangzhai83@gmail.com>

#include <vector>
#include <algorithm>
#include <stdio.h>
#include <stdlib.h>

typedef std::vector<int> RandNumList;

static void binary_search(RandNumList list, int low, int high, int key) 
{
    int mid = (high + low) / 2;
    if (low > high || mid <= low || mid >= high) 
        return;
    if (key == list[mid]) {
        printf("Found %d at %d\n", key, mid);
        return;
    }
    if (key < list[mid]) 
        binary_search(list, low, mid, key);
    else 
        binary_search(list, mid, high, key);
}

int main(int argc, char* argv[]) 
{
    RandNumList numList;
    for (int i = 0; i < 1024; i++) {
        numList.push_back(rand() % 100);
    }
    std::sort(numList.begin(), numList.end());
    printf("DEBUG: ");
    for (int i = 0; i < numList.size(); i++) {
        printf("%d\t", numList[i]);
    }
    printf("\n");
    binary_search(numList, 0, numList.size() - 1, argv[1] ? atoi(argv[1]) : 67);
    return 0;
}

File: src/search/binary_search.go
/* Copyright (C) 2013 ~ 2014 Leslie Zhai <xiangzhai83@gmail.com> */

package main

import (
    "fmt"
    "os"
    "strconv"
    "math/rand"
    "sort"
)

func binary_search(list []int, low, high, key int) {
    mid := (high + low) / 2
    if (low > high || mid <= low || mid >= high) { return }

    if (key == list[mid]) {
        fmt.Printf("Found %d at %d\n", key, mid)
        return
    }
    if (key < list[mid]) {
        binary_search(list, low, mid, key)
    } else {
        binary_search(list, mid, high, key)
    }
}

func main() {
    numList := []int{}
    for i := 0; i < 1024; i++ {
        numList = append(numList, rand.Intn(100))
    }
    sort.Ints(numList)
    fmt.Printf("DEBUG: ")
    for i := 0; i < len(numList); i++ {
        fmt.Printf("%d\t", numList[i])
    }
    fmt.Printf("\n")
    key := 67
    if len(os.Args) > 1 {
        key, _ = strconv.Atoi(os.Args[1])
    }
    fmt.Println("binary search", key)
    binary_search(numList, 0, len(numList) - 1, key)
}
